hal/network/ncp_client/sara/sara_ncp_client.cpp

```diff
@@ -1052,20 +1054,71 @@ int SaraNcpClient::selectSimCard(ModemState& state) {
 
     if (ncpId() == PLATFORM_NCP_SARA_R410) {
         int resetCount = 0;
+        // Note: Not failing on AT error on ICCID/IMSI lookup since SIMs have shown strange edge cases
+        // where they error for no reason, and hard resetting the modem or power cycling would not clear it.
+        //
         // Check if we are using a Twilio Super SIM based on ICCID
         char buf[32] = {};
-        auto lenCcid = getIccidImpl(buf, sizeof(buf));
-        CellularNetworkConfig netConfig;
-        CHECK_TRUE(lenCcid > 5, SYSTEM_ERROR_BAD_DATA);
-        netConf_ = networkConfigForIccid(buf, lenCcid);
+        int ccidCount = 0;
         do {
-            // Set UMNOPROF = SIM_SELECT
+            memset(buf, 0, sizeof(buf));
+            auto lenCcid = getIccidImpl(buf, sizeof(buf));
+            if (lenCcid > 5) {
+                netConf_ = networkConfigForIccid(buf, lenCcid);
```

Comment:
I noticed that we're also [overriding](https://github.com/particle-iot/device-os/blob/develop/hal/network/ncp_client/sara/sara_ncp_client.cpp#L1485) `netConf_` with the persistently stored (or default) settings in `configureApn()` which is called every time we instruct the client to connect to the network. The same logic with detection of the provider based on the IMSI or ICCID is duplicated there as well, although it only tried once. This duplication clearly needs some refactoring.

Perhaps as a quick fix we can prevent `configureApn()` from overriding `netConf_` if the configuration object passed to the method is empty? cc @avtolstoy

Reply:
Alternatively, let's store the provider settings in a local variable in `selectSimCard()` rather than in `netConf_`.

Reply:
From what I can see there is no reason in updating `netConf_` here, as it comes into play only when connecting, unless we want to skip detection logic in `configureApn()` if an empty configuration is passed as @sergeuz pointed out. Either solution looks fine :)

Reply:
It was suggested before to just simplify and use the global variable `netConf_` instead of the local one I had there :D

Reply:
I have the following suggestion:

- Let store the provider settings determined in `selectSimCard()` to a separate variable, e.g. `simNetConf_`.
- In `configureApn()` let's skip the detection if `conf` (the config passed to the method by the network manager) is empty and `simNetConf_` is not, and reuse `simNetConf_` for APN settings.
- Move the code that detects the provider settings based on IMSI and ICCID to a separate method and use that method in both `selectSimCard()` and `configureApn()` so that there's no code duplication.

I'll leave it up to you guys to decide whether it's something that we can do in this PR. I'm also fine with using a local variable in `selectSimCard()` as a quick fix.

Reply:
> Move the code that detects the provider settings based on IMSI and ICCID to a separate method and use that method in both selectSimCard() and configureApn() so that there's no code duplication

Actually, we can perform this detection just once in `selectSimCard()`, so there's probably no need to move it to a separate method.

Reply:
I'd like to keep the code the way it is currently.  I don't see it being duplicated since it's similar, yes... but it fails much differently.  The code in selectSimCard() to detect the SIM provider is designed to not fail completely in case of error, whereas the code in configureApn() is, and rightly so.  The "duplication" is a NIT we can ignore for now IMO.

Reply:
> I don't see it being duplicated since it's similar, yes... but it fails much differently

Given that the detection code is moved to a separate method, it's just a matter of how you're interpreting the return value of that method in `selectSimCard()` or `configureApn()`.

Reply:
I think we should table refactoring configureApn() code for optimization for another PR.  Minimize surface area of fixes for LTS.

Reply:
Sounds good. Please keep in mind that due to `netConf_` being set in multiple places it's generally not safe to rely on it to implement provider-specific workarounds. For example, [this workaround](https://github.com/particle-iot/device-os/blob/develop/hal/network/ncp_client/sara/sara_ncp_client.cpp#L1797-L1799) won't work if `netConf_` was overridden in `configureApn()`.

Reply:
I don't understand what you mean by "won't work if overridden".  We rely on configureApn() to set the netConf_ properly, so it better work :D

Reply:
We rely on `configureApn()` to set correct APN settings. The `CellularNetworkConfig` class didn't have a provider ID field previously (`netProv()`) – we added it at some point, but it's set to `NONE` in all objects passed to the NCP client by the network manager.

Reply:
When setting the correct APN via networkConfigForIccid()/networkConfigForImsi(), it also now sets the netProv(), so it should be fine, right?

Reply:
ah, you're right, we're performing the detection once again if the `conf` object is empty in `configureApn()`. I mixed up things, sorry. I guess the only issue then is that we're performing that detection twice, but as you've mentioned already, we can fix that in the scope of another PR :+1: